### PR TITLE
Improve config mapping for OpenTelemetry extensions

### PR DESCRIPTION
**Pull request by mcculls**
# What Does This Do

* Support arbitrary lookup of 'otel.' config from extensions
* Redirect InstrumentationConfig requests to our own ConfigProvider
* Redirect CallDepth requests to our own CallDepthThreadLocalMap
* Embed OTel instrumentation API and javaagent-extension API
* Remove classes covered by existing mappings as well as experimental packages
* Embed opentelemetry-javaagent-servlet-common-bootstrap
* Rework OTel package remapper to remap both shaded and unshaded references

Jira ticket: [APMAPI-64]

[APMAPI-64]: https://datadoghq.atlassian.net/browse/APMAPI-64?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-otel/otel-bootstrap/build.gradle [hunk 1]
```diff
@@ -33,8 +37,14 @@ spotbugs {
 
 dependencies {
   // latest OpenTelemetry API for drop-in support; instrumented at build-time with our shim
-  embeddedClasspath group: 'io.opentelemetry', name: 'opentelemetry-api', version: '1.38.0'
+  embeddedClasspath group: 'io.opentelemetry', name: 'opentelemetry-api', version: otelApiVersion
+  embeddedClasspath group: 'io.opentelemetry.instrumentation', name: 'opentelemetry-instrumentation-api', version: otelInstrumentationApiVersion
+  embeddedClasspath group: 'io.opentelemetry.javaagent', name: 'opentelemetry-javaagent-extension-api', version: "$otelInstrumentationApiVersion-alpha"
+
+  // selected bootstrap types shared across multiple OpenTelemetry instrumentations
+  embeddedClasspath group: 'io.opentelemetry.javaagent.instrumentation', name: 'opentelemetry-javaagent-servlet-common-bootstrap', version: "$otelInstrumentationApiVersion-alpha"
```
Comment (PerfectSlayer) [inline dd-java-agent/agent-otel/otel-bootstrap/build.gradle:45]: Just checking: can this cause an issue with customers shipping OTel SDK (especially another version) as part of their dependencies?

**Review by PerfectSlayer — APPROVED** [review]
Sounds good.
One quick check in the comments.

Otherwise, OTel supports Spring as config source.
Can it cause troubles for Spring customer if we don't? Do we plan to support it too?

File: dd-java-agent/agent-otel/otel-bootstrap/build.gradle (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/agent-otel/otel-bootstrap/build.gradle:45]: no, the embedded classpath is shaded - see https://github.com/DataDog/dd-trace-java/blob/mcculls/map-otel-config-for-dropin/dd-java-agent/agent-otel/otel-bootstrap/build.gradle#L97

**Comment by mcculls** [comment]
> Otherwise, OTel supports Spring as config source. Can it cause troubles for Spring customer if we don't? Do we plan to support it too?

That's outside the scope of this feature - the focus here is allowing the same config provider (system properties, environment variables, config file) to feed configuration to both Datadog and OpenTelemetry instrumentations.

File: dd-java-agent/agent-otel/otel-bootstrap/build.gradle (hunk 1 above)
Comment (PerfectSlayer) [reply dd-java-agent/agent-otel/otel-bootstrap/build.gradle:45]: Between custom Gradle configuration, shadowing and `ClassRemapper` implementation, I wonder if there is a better way (in term of maintainability) to express the single need of "I want this dependency under this package" 🤔

**Comment by PerfectSlayer** [comment]
> If/when we add the ability to configure the Datadog tracer from Spring then this PR would let that configuration flow through to both Datadog and OpenTelemetry instrumentations.

Sure, it was more to enquire about the next roadmap 😇

File: dd-java-agent/agent-otel/otel-bootstrap/build.gradle (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/agent-otel/otel-bootstrap/build.gradle:45]: You could look at using a single configuration file - the downside is that you'd need more code on both the gradle and Java sides to load that file into the respective plugins, increasing their complexity. Sometimes duplication is acceptable for things that change infrequently...

**Pull request merged**